### PR TITLE
=act #21237 fix regression in ByteString.slice

### PR DESCRIPTION
**Pull request by ktoso**
Submitting drop().take() version of https://github.com/akka/akka/pull/21248 
Attributing commit to @monkey-mas, thanks for helping us out here - and looking forward to the better optimised slice() that passes all crazy edge cases!

### CODE REVIEW

**Comment by ktoso** [comment]
@akka/akka-team this can be merged once passes validation, it is the simplified version of #21248 

Resolves  #21237

**Comment by ktoso** [comment]
!! please make sure to merge as merge commit, so @monkey-mas get's attributed !!

**Comment by patriknw** [comment]
LGTM

File: akka-actor/src/main/scala/akka/util/ByteString.scala [hunk 1]
```diff
@@ -164,7 +164,8 @@ object ByteString {
 
     override def slice(from: Int, until: Int): ByteString =
       if ((from == 0) && (until == length)) this
-      else if (from > length) ByteString.empty
+      else if (from >= length) ByteString.empty
+      else if (until <= 0) ByteString.empty
```
Comment (drewhk) [inline akka-actor/src/main/scala/akka/util/ByteString.scala]: so until is not inclusive, right?

**Comment by drewhk** [comment]
LGTM

File: akka-actor-tests/src/test/scala/akka/util/ByteStringSpec.scala [hunk 2]
```diff
@@ -364,14 +361,49 @@ class ByteStringSpec extends WordSpec with Matchers with Checkers {
       ByteStrings(ByteString1.fromString("ab"), ByteString1.fromString("")).drop(Int.MinValue) should ===(ByteString("ab"))
       ByteStrings(ByteString1.fromString(""), ByteString1.fromString("ab")).dropRight(Int.MinValue) should ===(ByteString("ab"))
     }
-    "slice" in {
-      ByteStrings(ByteString1.fromString("a"), ByteString1.fromString("")).slice(0, 1) should ===(ByteString("a"))
+    "slice " in {
       ByteStrings(ByteString1.fromString(""), ByteString1.fromString("a")).slice(1, 1) should ===(ByteString(""))
+      // Can obtain expected results from 6 basic patterns
```
Comment (viktorklang) [inline akka-actor-tests/src/test/scala/akka/util/ByteStringSpec.scala]: Doesn't this look like a perfect candidate for generative testing with ScalaCheck?

File: akka-actor/src/main/scala/akka/util/ByteString.scala (hunk 1 above)
Comment (viktorklang) [inline akka-actor/src/main/scala/akka/util/ByteString.scala]: `else if (from >= length || until <= 0) ByteString.empty` ?

File: akka-actor/src/main/scala/akka/util/ByteString.scala [hunk 3]
```diff
@@ -252,11 +253,8 @@ object ByteString {
       if (n <= 0) ByteString.empty
       else ByteString1(bytes, startIndex, Math.min(n, length))
 
-    override def slice(from: Int, until: Int): ByteString = {
-      if (from <= 0 && until >= length) this // we can do < / > since we're Compact
-      else if (until <= from) ByteString1.empty
-      else ByteString1(bytes, startIndex + from, until - from)
-    }
+    override def slice(from: Int, until: Int): ByteString =
+      drop(from).take(until - Math.max(0, from))
```
Comment (viktorklang) [inline akka-actor/src/main/scala/akka/util/ByteString.scala]: What's the performance implication of the change?

File: akka-actor/src/main/scala/akka/util/ByteString.scala [hunk 4]
```diff
@@ -164,7 +164,8 @@ object ByteString {
 
     override def slice(from: Int, until: Int): ByteString =
       if ((from == 0) && (until == length)) this
```
Comment (2beaucoup) [inline akka-actor/src/main/scala/akka/util/ByteString.scala]:
```scala
if (from <= 0 && until >= length) this
```

**Comment by 2m** [comment]
LGTM

File: akka-actor/src/main/scala/akka/util/ByteString.scala (hunk 3 above)
Comment (patriknw) [inline akka-actor/src/main/scala/akka/util/ByteString.scala]: it was discussed in the other PR: https://github.com/akka/akka/pull/21248#discussion_r76156413

File: akka-actor/src/main/scala/akka/util/ByteString.scala (hunk 1 above)
Comment (2beaucoup) [inline akka-actor/src/main/scala/akka/util/ByteString.scala]:
```scala
else if (from >= length || until <= 0 || from >= until) ByteString.empty
```
?

File: akka-actor-tests/src/test/scala/akka/util/ByteStringSpec.scala (hunk 2 above)
Comment (drewhk) [inline akka-actor-tests/src/test/scala/akka/util/ByteStringSpec.scala]: I was tested via ScalaCheck but it did not uncover the issue. But @ktoso has more information on this.

Comment (ktoso) [inline akka-actor-tests/src/test/scala/akka/util/ByteStringSpec.scala]: We have ScalaChecks, and if it finds an edge case that I for sure what to cover,
that totally makes sense to put into an explicit test. Esp. because this is exactly how we missed this bug: "oh yeah we have scalacheck!" which gave us false confidence.

Only after running 1000s times (instead of default settings) ScalaCheck would get into these edge cases. So I want them to be specifically covered here, in addition to SC.

Comment (viktorklang) [inline akka-actor-tests/src/test/scala/akka/util/ByteStringSpec.scala]: You're absolutely right, would be nice to document these as regression tests :)

Comment (ktoso) [inline akka-actor-tests/src/test/scala/akka/util/ByteStringSpec.scala]: You mean with a ref to ticket? Sure, can add 👍

File: akka-actor/src/main/scala/akka/util/ByteString.scala (hunk 4 above)
Comment (ktoso) [inline akka-actor/src/main/scala/akka/util/ByteString.scala]: goodtrue

**Comment by ktoso** [comment]
Updated

File: akka-actor-tests/src/test/scala/akka/util/ByteStringSpec.scala [hunk 5]
```diff
@@ -22,10 +22,7 @@ import scala.collection.mutable.Builder
 class ByteStringSpec extends WordSpec with Matchers with Checkers {
 
   //  // uncomment when developing locally to get better coverage
-  //  implicit override val generatorDrivenConfig =
-  //    PropertyCheckConfig(
-  //      minSuccessful = 1000,
-  //      minSize = 0, maxSize = 100)
+  implicit val betterGeneratorDrivenConfig = PropertyCheckConfig().copy(minSuccessful = 1000)
```
Comment (johanandren) [inline akka-actor-tests/src/test/scala/akka/util/ByteStringSpec.scala]: should this be left uncommented? if so, remove the comment above it

File: akka-actor-tests/src/test/scala/akka/util/ByteStringSpec.scala [hunk 6]
```diff
@@ -365,13 +362,48 @@ class ByteStringSpec extends WordSpec with Matchers with Checkers {
       ByteStrings(ByteString1.fromString(""), ByteString1.fromString("ab")).dropRight(Int.MinValue) should ===(ByteString("ab"))
     }
     "slice" in {
-      ByteStrings(ByteString1.fromString("a"), ByteString1.fromString("")).slice(0, 1) should ===(ByteString("a"))
       ByteStrings(ByteString1.fromString(""), ByteString1.fromString("a")).slice(1, 1) should ===(ByteString(""))
+      // We explicitly test all edge cases to always test them, refs bug #21237
```
Comment (johanandren) [inline akka-actor-tests/src/test/scala/akka/util/ByteStringSpec.scala]: good!

**Comment by johanandren** [comment]
LGTM

File: akka-actor-tests/src/test/scala/akka/util/ByteStringSpec.scala (hunk 5 above)
Comment (ktoso) [inline akka-actor-tests/src/test/scala/akka/util/ByteStringSpec.scala]: I'll comment uncomment it perhaps, good to have more coverage here
